txnsync/txngroups.go

```diff
@@ -60,23 +69,61 @@ func encodeTransactionGroups(inTxnGroups []transactions.SignedTxGroup) ([]byte,
 					stub.BitmaskGroup.SetBit(index)
 				}
 				if err := stub.deconstructSignedTransactions(index, &txn); err != nil {
-					return nil, fmt.Errorf("failed to encodeTransactionGroups: %w", err)
+					return nil, 0, fmt.Errorf("failed to encodeTransactionGroups: %w", err)
 				}
 				index++
 			}
 		}
 	}
 	stub.finishDeconstructSignedTransactions()
 
-	return stub.MarshalMsg(protocol.GetEncodingBuf()[:0]), nil
+	encoded := stub.MarshalMsg(protocol.GetEncodingBuf()[:0])
+
+	// check if time saved by compression: estimatedGzipCompressionGains * len(msg) / dataExchangeRate
+	// is greater than by time spent during compression: len(msg) / estimatedGzipCompressionSpeed
+	if len(encoded) > minEncodedTransactionGroupsCompressionThreshold && float32(dataExchangeRate) < (estimatedGzipCompressionGains * estimatedGzipCompressionSpeed) {
+		compressedBytes, err := compressTransactionGroupsBytes(encoded)
+		if err == nil {
+			return compressedBytes, 1, nil
+		}
+		logging.Base().Warnf("failed to compress txnsync msg: %v", err)
```

Comment:
make the `encodeTransactionGroups` method a `(s *syncState)` method, and use the `s.log` to log warning messages. Also, when reporting the error, make sure to report the size of the input buffer.